### PR TITLE
Suggest unwrap_or_else when a closure is given

### LINKED ISSUE #102320: Suggest `unwrap_or_else` when a closure is passed to `unwrap_or`  instead of suggesting calling it
`rustc` should suggest `unwrap_or_else` if the argument is a closure instead of suggesting to call the closure

Given the following code:  https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=15e99f8e75edf68460c457c09e4e2fa4

```rust
fn main() {
  let x = "com.example.app";
  let y: Option<&str> = None;
  let s = y.unwrap_or(|| x.split('.').nth(1).unwrap());
}
```

The current output is:

```
error[[E0308]](https://doc.rust-lang.org/stable/error-index.html#E0308): mismatched types
 --> src/main.rs:4:23
  |
4 |   let s = y.unwrap_or(|| x.split('.').nth(1).unwrap());
  |             --------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&str`, found closure
  |             |
  |             arguments to this function are incorrect
  |
  = note: expected reference `&str`
               found closure `[closure@src/main.rs:4:23: 4:25]`
note: associated function defined here
help: use parentheses to call this closure
  |
4 |   let s = y.unwrap_or((|| x.split('.').nth(1).unwrap())());
  |                       +                               +++
```

Ideally the output should look like:

```
error[[E0308]](https://doc.rust-lang.org/stable/error-index.html#E0308): mismatched types
 --> src/main.rs:4:23
  |
4 |   let s = y.unwrap_or(|| x.split('.').nth(1).unwrap());
  |             --------- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&str`, found closure
  |             |
  |             arguments to this function are incorrect
  |
  = note: expected reference `&str`
               found closure `[closure@src/main.rs:4:23: 4:25]`
note: associated function defined here
help: use "unwrap_or_else"
  |
4 |   let s = y.unwrap_or_else(|| x.split('.').nth(1).unwrap());
  |                      +++++                              
```

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #102320

r? @compiler-errors

### CODE REVIEW

**Comment by chenyukang** [comment]
May need more work to expand to suggest `map_or_else`, etc.

File: compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs [hunk 1]
```diff
@@ -80,6 +80,38 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+            // suggest `unwrap_or_else` for `Option` and `Result`, when a closure is given
+            if let hir::ExprKind::Closure(..) = expr.kind &&
+                let parent = self.tcx.hir().get_parent_node(expr.hir_id) &&
+                let Some(Node::Expr(call_expr)) = self.tcx.hir().find(parent) &&
+                let hir::ExprKind::MethodCall(
+                      hir::PathSegment {ident: method_name, ..},
+                      receiver,
+                      _args,
+                      _span) = call_expr.kind &&
+                inputs.len() == 0 {
```
Comment (chenyukang) [inline compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs:92]: I don't know how to check the arguments types, currently using `method_exists` to check the function `unwrap_or_else` existed or not.

File: compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs [hunk 2]
```diff
@@ -80,6 +80,38 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let Some((def_id_or_name, output, inputs)) = self.extract_callable_info(expr, found)
             else { return false; };
         if can_satisfy(output) {
+            // suggest `unwrap_or_else` for `Option` and `Result`, when a closure is given
+            if let hir::ExprKind::Closure(..) = expr.kind &&
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs:84]: If you use `extract_callable_info` then you don't need to also check if the expr is a closure.

File: compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs:92]: Please check that the `_else` method takes the same number of arguments at least

File: compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs [hunk 3]
```diff
@@ -80,6 +80,38 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                      _args,
+                      _span) = call_expr.kind &&
+                inputs.len() == 0 {
+              let self_ty = self.typeck_results.borrow().expr_ty(receiver);
+              let new_name = Ident {
+                  name: Symbol::intern(&format!("{}_else", method_name.as_str())),
+                  span: method_name.span,
+              };
+
+              if self.method_exists(
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs:99]: I would use `lookup_probe` instead, so you can give it this probe scope: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir_analysis/check/method/probe/enum.ProbeScope.html#variant.TraitsInScope so we don't suggest methods that are out of scope

File: compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs [hunk 4]
```diff
@@ -80,6 +80,38 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                  new_name,
+                  self_ty,
+                  call_expr.hir_id,
+                  false,
+              ) {
+                  err.span_suggestion_verbose(
+                      method_name.span.shrink_to_hi(),
+                      &format!("try calling `{}` instead", new_name.name.as_str()),
+                      "_else",
+                      Applicability::MachineApplicable,
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs:109]: This isn't machine applicable...? It's very possible for the suggestion to be wrong :sweat_smile:

File: compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs [hunk 5]
```diff
@@ -80,6 +80,38 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let Some((def_id_or_name, output, inputs)) = self.extract_callable_info(expr, found)
             else { return false; };
         if can_satisfy(output) {
+            // suggest `unwrap_or_else` for `Option` and `Result`, when a closure is given
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs:83]: Please move this logic to a different function. It doesn't make sense for this to live in `suggest_fn_call` -- since this isn't suggesting a fn call.

**Review by compiler-errors — CHANGES_REQUESTED** [review]

File: compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs (hunk 4 above)
Comment (chenyukang) [reply compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs]: Fixed!

File: compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs (hunk 5 above)
Comment (chenyukang) [reply compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs]: Moved to function `suggest_else_fn_with_closure`.

File: compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs (hunk 3 above)
Comment (chenyukang) [reply compiler/rustc_hir_analysis/src/check/fn_ctxt/suggestions.rs]: `method_exists` also use `TraitsInScope`, https://github.com/rust-lang/rust/blob/ab37a83ae2c829b8597bdcda8244964ebc60b495/compiler/rustc_hir_analysis/src/check/method/mod.rs#L104

I changed to use `lookup_probe` to get the pick result, and check the function arguments..

File: compiler/rustc_hir_analysis/src/check/method/suggest.rs [hunk 6]
```diff
@@ -2324,6 +2324,61 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
     }
 
+    /// issue #102320, for `unwrap_or` with closure as argument, suggest `unwrap_or_else`
+    /// FIXME: currently not working for suggesting `map_or_else`, see #102408
+    pub(crate) fn suggest_else_fn_with_closure(
+        &self,
+        err: &mut Diagnostic,
+        expr: &hir::Expr<'_>,
+        found: Ty<'tcx>,
+        can_satisfy: impl FnOnce(Ty<'tcx>) -> bool,
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/check/method/suggest.rs:2334]: This doesn't need to be an argument. Can you inline the `can_coerce` call here by passing in `expected: Ty<'tcx>`?

File: compiler/rustc_hir_analysis/src/check/method/suggest.rs [hunk 7]
```diff
@@ -2324,6 +2324,61 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+        let parent = self.tcx.hir().get_parent_node(expr.hir_id);
+        if  let Some(Node::Expr(call_expr)) = self.tcx.hir().find(parent) &&
+            let hir::ExprKind::MethodCall(
+            hir::PathSegment { ident: method_name, .. },
+            self_expr,
+            args,
+            ..,
+        ) = call_expr.kind
+        {
+            let self_ty = self.typeck_results.borrow().expr_ty(self_expr);
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/check/method/suggest.rs:2351]: I think there is an `expr_ty_opt` version of this that doesn't ICE if the node is missing a type If so, then please use that, just in case.

File: compiler/rustc_hir_analysis/src/check/method/suggest.rs [hunk 8]
```diff
@@ -2324,6 +2324,61 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                self_expr,
+                ProbeScope::TraitsInScope,
+            );
+
+            // check the method arguments number
+            if let Ok(pick) = probe &&
+                let fn_ty = self.tcx.type_of(pick.item.def_id) &&
+                let fn_sig = fn_ty.fn_sig(self.tcx) &&
+                let fn_args = fn_sig.skip_binder().inputs() &&
+                fn_args.len() == args.len() +1 {
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/check/method/suggest.rs:2369]:
```suggestion
                fn_args.len() == args.len() + 1 {
```

File: compiler/rustc_hir_analysis/src/check/method/suggest.rs [hunk 9]
```diff
@@ -2324,6 +2324,61 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                new_name,
+                self_ty,
+                self_expr,
+                ProbeScope::TraitsInScope,
+            );
+
+            // check the method arguments number
+            if let Ok(pick) = probe &&
+                let fn_ty = self.tcx.type_of(pick.item.def_id) &&
+                let fn_sig = fn_ty.fn_sig(self.tcx) &&
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/check/method/suggest.rs:2367]:
```suggestion
                let fn_sig = self.tcx.fn_sig(pick.item.def_id) &&
```

**Comment by compiler-errors** [comment]
Few more nits

**Comment by chenyukang** [comment]
> Few more nits

Fixed, thanks!

**Comment by compiler-errors** [comment]
Hm, this suggestion might have some false positives still, but this is probably ok...

@bors r+
